File: android/Pets/app/src/main/java/com/pets/domain/favorites/FavoritesReducer.kt
package com.pets.domain.favorites

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import com.pets.architecture.common.Depending
import com.pets.architecture.reducers.Reducing
import com.pets.domain.list.ListEvent
import com.pets.models.Pet
import com.pets.services.ServiceLocator
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

class FavoritesReducer: Reducing<FavoritesState, FavoritesEvent>,
    Depending<FavoritesReducer.Dependencies> {
    class Dependencies {
        var service = ServiceLocator.favoriteService
        var logService = ServiceLocator.logService
    }

    override val dependencies: Dependencies
        get() = Dependencies()

    override suspend fun reduce(state: MutableLiveData<FavoritesState>, event: FavoritesEvent): FavoritesEvent?
    = withContext(Dispatchers.IO) {
        when(event) {
            is RemoveFromFavorites -> {
                state.value?.let { fs ->
                    var favorites = fs.list.toMutableList()
                    favorites.removeAll { it.id == event.id }
                    state.postValue(fs.copy(list = favorites))
                }
            }

            is FetchListCompleted -> {
                state.value?.let { fs ->
                    state.postValue(fs.copy(list = event.list,
                    isLoading = false))
                }
            }

            is FetchList -> {
                state.value?.let {
                    state.postValue(it.copy(list = emptyList(), isLoading = true))
                }
            }
        }
        return@withContext sideEffect(event = event)
    }

    private suspend fun sideEffect(event: FavoritesEvent): FavoritesEvent? {
        dependencies.logService.log(event)
        when (event) {
            is FetchList -> {
                val list = dependencies.service.fetchList()
                return FetchListCompleted(list = list)
            }

            is RemoveFromFavorites -> {
                dependencies.service.removeFromFavorites(id = event.id)
            }

            else -> return null
        }

        return null
    }
}

File: ios/Pets/Domain/Sources/Domain/Favorites/Reducers/FavoritesReducer.swift
//
//  File.swift
//  
//
//  Created by Dylan Velez on 10/08/21.
//
import Foundation
import Architecture
import Services
import Combine

public struct FavoritesReducer: Reducing, Depending {
    public typealias State = FavoritesState
    public typealias Event = FavoritesEvent
    
    public struct Dependencies {
        public var service: FavoriteService = ServiceLocator.favoriteService
        public var logService: LoggingService = ServiceLocator.logService
        public init(service: FavoriteService, logService: LoggingService) {
            self.service = service
            self.logService = logService
        }
    }
    
    public var dependencies: Dependencies
    
    public init(service: FavoriteService, logService: LoggingService) {
        self.dependencies = Dependencies(service: service, logService: logService)
    }
    
    public func reduce(state: inout FavoritesState, forEvent event: FavoritesEvent) -> Effect<Event> {
        
        switch event {
        case .removeFromFavorites(let id):
            state.list.removeAll(where: { $0.id == id })
        case .fetchListCompleted(let list):
            state.list = list
            state.isLoading = false
        case .fetchList:
            state.isLoading = true
            break
        }
        
        return sideEffect(event: event)
    }
    
    private func sideEffect(event: Event) -> Effect<Event> {
        dependencies.logService.log(action: event)
        switch event {
        case .fetchList:
            return dependencies.service.fetchList()
                .map(FavoritesEvent.fetchListCompleted)
                .catch { error in
                    return Just(.fetchListCompleted(pets: []))
                }
                .eraseToAnyPublisher()
            
        case .removeFromFavorites(let id):
            dependencies.service.removeFromFavorites(id: id)
        case .fetchListCompleted:
            break
        }
        return nil
    }
}
